# `📅DAY_04` _기초 AI Programming

## 데이터 분석과 Pandas

### Pandas

- 파이썬에서 데이터를 다루기 쉽게 만들어주는 라이브러리
- 주로 표(엑셀처럼 행과 열이 있는 데이터)를 다룰 때 사용
- 현재는 **데이터 과학, 인공지능(AI), 금융 분석 등 다양한 분야**에서 널리 사용됨.
- Pandas의 **DataFrame**라는 기능 : **엑셀 표처럼 데이터를 정리해주는 테이블**

#### Pandas를 사용하는 이유

1. 데이터를 가져오는 기능
- Pandas를 사용하면 파일(엑셀, CSV, 데이터베이스 등)에서 데이터를 쉽게 불러올 수 있음.

2. 데이터 정리 기능
- 불필요한 데이터는 버리고, 필요한 데이터만 정리해서 사용

3. 데이터 변형 및 분석
- 데이터를 변형하고, 가공하고, 분석하는 과정

4. 데이터 시각화
- 데이터를 그래프나 표로 시각화하여 쉽게 이해할 수 있음.


### Excel과 Pandas

Python과 Pandas를 사용하면 **더 빠르고, 정확하며, 효율적**으로 데이터 분석을 할 수 있음.

**엑셀**과 비교하여 설명해보자면,
1) 반복적인 작업을 자동화할 수 있다.
    - 한 번 작성한 코드로 수천 개의 파일도 자동으로 처리 가능

2) 대용량 데이터를 빠르게 처리할 수 있다.

3) 데이터를 더 정확하게 분석할 수 있다.
    - 코드가 정확하게 실행되기 때문에 실수를 줄일 수 있음.

4) 데이터를 쉽게 변형하고 분석할 수 있다.
    - 한 줄의 코드로 쉽게 데이터를 정리할 수 있음.

5) 그래프와 보고서를 자동으로 만들 수 있다.
    - **Matplotlib, Seaborn** 같은 라이브러리를 사용하면 데이터가 변해도 자동으로 그래프를 생성할 수 있음.
    - **새로운 데이터가 추가**되더라도 자동으로 보고서를 생성할 수 있음.


### Series와 DataFrame

#### Series

- 하나의 열(Column)을 나타냄.
- **같은 종류의 데이터**를 모아둔 것.

In [3]:
# Series 예제

# pandas 라이브러리 불러오기
import pandas as pd

# 학생들의 나이를 Series로 만들기 (과일 상자처럼 한 가지 데이터만 저장)
# pd.Series([...])는 Series를 생성하는 명령어
ages = pd.Series([25, 30, 22, 28, 35])

print(ages)

0    25
1    30
2    22
3    28
4    35
dtype: int64


#### DataFrame

- 엑셀의 표(테이블)와 같은 데이터 구조
- 여러 개의 Series(열)이 모여서 만들어진 형태

**리스트를 이용한 데이터프레임 생성**
- 리스트를 이용하면 **한 줄(행)씩 데이터를 저장**할 수 있음.
- **각 행(가로)** = 하나의 리스트로 저장
- **각 열(세로)** = 데이터프레임에서 자동 정렬

In [7]:
# 리스트를 이용한 데이터프레임 생성

import pandas as pd

# 학생 정보를 리스트로 저장 (각 행을 리스트로 표현)
# data는 리스트 안에 또 다른 리스트가 들어 있는 구조
# 각 리스트는 하나의 행(Row)을 의미

data = [
    ["김철수", 25, 175],
    ["이영희", 30, 162],
    ["박지민", 22, 168]
]

# pd.DataFrame(data) -> 리스트를 데이터프레임으로 변환
df = pd.DataFrame(data, columns=['학생 이름', '나이', '키'])
df

,학생 이름,나이,키
0,김철수,25,175
1,이영희,30,162
2,박지민,22,168


In [6]:
# 딕셔너리를 이용해 데이터프레임 생성하기

import pandas as pd

data = {
    "학생 이름": ['김철수', '이영희', '박지민'],
    '나이': [25, 30, 22],
    '키': [175, 162, 168]
}

df = pd.DataFrame(data)
df

,학생 이름,나이,키
0,김철수,25,175
1,이영희,30,162
2,박지민,22,168


## 반도체 공정과 데이터

### 반도체 공정 개요

#### 반도체 제조 공정

1. 웨이퍼 제조
- 반도체를 만들려면 실리콘 웨이퍼가 필요함.

2. 산화 공정(Oxidation)
- 웨이퍼 표면을 보호하기 위해 **산화막**(SiO2)을 씌움.
- 이 산화막은 반도체가 잘 만들어질 수 있도록 돕고, 이후의 공정에서 원하는 패턴을 만들 수 있도록 함.

3. 포토 공정(Photolithography)
- 빛을 이용해 **특정 부분에만 원하는 회로 모양(=패턴)을 새김.**

4. 식각 공정(Etching)
- **필요 없는 부분을 제거**하여 회로의 윤곽을 만듦.

5. 증착/이온주입 공정
- 1) 증착 공정(Deposition)
    - 반도체 표면에 **특정한 막을 추가하는 과정**
    - 전기 신호를 잘 전달하도록 도와주거나, 보호하는 역할
- 2) 이온 주입 공정(Ion Implantation)
    - 반도체 내부에서 전기가 더 잘 흐르거나, 특정 부분에서만 전기가 통하게 만드는 과정
    - 이온을 주입하면 반도체가 특정한 성질(전기가 잘 흐르거나 막히는 성질)을 가질 수 있음.

6. 금속배선 공정
- 금속을 이용해 **회로 간의 연결을 만들어 금속 배선을 통해 전기 신호가 흐를 수 있도록** 만듦.

7. EDS 공정
- **전기적 특성 검사(EDS)** 를 통해 각 반도체 칩이 정상적으로 동작하는지 확인.
- 품질 검사를 통해 정상적인 제품만 선별

8. 패키징 공정
- 칩을 포장하여 다양한 기기에 사용할 수 있도록 함.

### 반도체 산화 공정 데이터

#### 반도체 산화 공정

- **Wafer**라는 실리콘 원판 위에 일정한 두께의 산화막을 형성하는 과정

**산화막(SiO2)의 역할**
- 절연층 역할 : 전기가 흐르지 않아야 되는 부분에 전기가 못 흐르게 막는 역할
- 산소나 불순물 차단 : 웨이퍼가 외부 오염에 영향을 받지 않도록 방어
- 게이트 산화막 : 산화막은 얇은 유리처럼 생긴 막으로, 전기가 새거나 섞이지 않도록 막아주는 역할을 함.

#### 1. 건식 산화 (Dry Oxidation)

- **얇고 균일한 고품질의 산화막을 형성 (절연층 역할)**
    - 절연층은 반도체 소자 내에서 전기가 흐르지 않도록 막는층, 즉 전기적으로 서로 간섭하지 않도록 보호해주는 역할을 하는 층으로 산화막이 절연층의 역할을 함.
- **산화 속도가 느림** (습식 산화보다 시간이 오래 걸림)
- **고온(900~1100도)에서 순수 산소(O2)만 사용하여 반응**

- 트랜지스터의 **게이트 산화막 형성**에 사용됨 (핵심 반도체 소자에 활용)
    - 트랜지스터는 전기를 켜고 끄는 전자 스위치 역할을 함.
    - 트랜지스터에는 Gate를 통해 전류 흐름을 제어함.
    - 건식 산화 공정은 이 Gate 산화막을 만드는 데 매우 중요한 공정


#### 2. 습식 산화 (Wet Oxidation)

- **수증기(H2O)와 산소(O2)** 를 함께 사용해 산화막 형성
- 두꺼운 산화막을 빠르게 형성할 수 있지만, 건식 산화보다 품질이 다소 떨어질 수 있음.


**산화 공정에서 중요한 변수 -> 온도, 시간, 산소 농도, 압력**
- 이 변수들이 산화막의 두께와 균일성에 영향을 미친다.

#### 컬럼 정보

**1. 메타 정보**

공정을 언제, 어떤 장비에서 수행했는지를 나타냄.

- **process_date**: 공정을 진행한 날짜
- **process_time**: 공정을 진행한 시간
- **equipment_id**: 사용된 반도체 제조 장비의 ID (어떤 장비에서 공정을 수행했는지 기록)
- **chamber_id**: 공정이 진행된 챔버(산화 반응이 일어나는 공간)의 ID


**2. 건식 산화 공정 진행 중 (During Dry Oxidation Process)**

웨이퍼를 챔버(반응기)에 넣고 산화막을 형성하는 과정을 기록한 데이터

- **wafer_id**: 가공된 웨이퍼의 개별 식별 번호
- **wafer_doping_type**: 웨이퍼가 전기를 잘 통하게 하기 위해 첨가한 물질의 종류 (P형, N형)
- **precleaning_time**: 산화 전에 웨이퍼를 세척하는 데 걸린 시간 (단위: **초(sec)**)
- **oxidation_temperature**: 웨이퍼가 산화되는 동안 가열된 온도 (단위: **°C**)
- **oxidation_time**: 산화 공정이 진행된 시간 (단위: **분**)
- **pressure**: 공정 중 챔버 내부의 압력 (단위: **Torr**)
- **gas_flow_rate**: 공정 중 웨이퍼 위로 흘려보낸 산소 또는 기타 가스의 양 (단위: **sccm**, 즉 cc/min)


**3. 공정이 끝난 후 (After Dry Oxidation Process)**

산화 공정이 완료된 후 웨이퍼의 품질을 확인하는 과정을 기록한 데이터

- **oxide_thickness**: 웨이퍼 표면에 형성된 산화막의 두께 (단위: **나노미터 nm**)
- **uniformity**: 산화막이 얼마나 균일하게 형성되었는지를 백분율(%)로 나타낸 값
    - 웨이퍼의 여러 지점에서 산화막의 두께를 측정하여 균일성을 확인
    - 1-Sigma 균일성(%) 공식:
    ```bash
    1-Sigma 균일성 = (σ/μ) x 100
    ```
    - 1-Sigma 균일성 값이 **낮을수록 오차가 적고, 모든 지점에서 거의 같은 두께를 유지하고 있다**는 의미

- **defect_count**: 웨이퍼 표면에서 발견된 결함(불량) 개수
- **defect_type**: 검출된 결함(불량)의 유형
    |**결함 유형**|**설명**|**후처리 가능 여부**|**사용 가능 여부**|
    |---|---|---|---|
    |none|결함 없음|후처리 불필요|사용 가능|
    |particle|표면 먼지, 입자|세척 후 가능|심할 경우 폐기|
    |pinhole|미세한 구멍|작으면 보완 가능|깊거나 크면 폐기|
    |crack|균열, 깨짐|후처리 불가능|폐기|

## 데이터 분석의 기초

### CSV 파일 읽기

- **CSV** : **쉼표(,)** 로 구분된 텍스트 파일

In [8]:
# 산화 공정 데이터 읽어서 화면에 출력

import pandas as pd

# file_path는 불러올 CSV 파일의 경로(파일명)
file_path = "02_oxide.csv"

# pd.read_csv(file_path)를 사용하면 CSV 파일이 '데이터프레임'으로 변환됨.
# oxide_df에 CSV 파일 데이터 저장
oxide_df = pd.read_csv(file_path)

oxide_df

,process_date,process_time,equipment_id,chamber_id,wafer_id,wafer_doping_type,precleaning_time,oxidation_temperature,oxidation_time,pressure,gas_flow_rate,oxide_thickness,uniformity,defect_count,defect_type
0,2025-01-01,0:00:00,eq_1,chamber_1,w_0,p-type,3,908.332942,30.450712,0.725458,4007.192654,212.318551,4.158928,0,none
1,2025-01-01,1:00:00,eq_2,chamber_2,w_1,n-type,1,998.595807,43.926035,0.610652,3273.074465,359.152236,3.920767,2,particle
2,2025-01-01,2:00:00,eq_3,chamber_3,w_2,p-type,5,1016.761934,66.715328,0.816619,2098.663479,442.100760,2.428848,1,pinhole
3,2025-01-01,3:00:00,eq_4,chamber_1,w_3,n-type,3,952.625393,46.845448,0.761523,2897.739657,225.153908,0.437444,2,particle
4,2025-01-01,4:00:00,eq_5,chamber_2,w_4,p-type,3,968.690498,52.487699,0.649097,2335.106388,423.234943,2.860362,0,none
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,2025-02-11,11:00:00,eq_1,chamber_3,w_995,n-type,4,1068.991935,52.783496,0.302645,4802.014246,464.065417,4.914362,0,none
996,2025-02-11,12:00:00,eq_2,chamber_1,w_996,p-type,5,976.739157,55.965298,0.696086,2534.436161,261.114871,3.471068,0,none
997,2025-02-11,13:00:00,eq_3,chamber_2,w_997,n-type,3,985.560235,45.612643,0.648714,2697.488104,310.582879,0.294383,0,none
998,2025-02-11,14:00:00,eq_4,chamber_3,w_998,p-type,2,946.535657,6.432315,0.923589,3858.784356,179.629674,2.242628,4,particle


In [9]:
# 데이터프레임 출력

# .head()는 데이터의 처음 5개 행을 출력
oxide_df.head()

,process_date,process_time,equipment_id,chamber_id,wafer_id,wafer_doping_type,precleaning_time,oxidation_temperature,oxidation_time,pressure,gas_flow_rate,oxide_thickness,uniformity,defect_count,defect_type
0,2025-01-01,0:00:00,eq_1,chamber_1,w_0,p-type,3,908.332942,30.450712,0.725458,4007.192654,212.318551,4.158928,0,none
1,2025-01-01,1:00:00,eq_2,chamber_2,w_1,n-type,1,998.595807,43.926035,0.610652,3273.074465,359.152236,3.920767,2,particle
2,2025-01-01,2:00:00,eq_3,chamber_3,w_2,p-type,5,1016.761934,66.715328,0.816619,2098.663479,442.100760,2.428848,1,pinhole
3,2025-01-01,3:00:00,eq_4,chamber_1,w_3,n-type,3,952.625393,46.845448,0.761523,2897.739657,225.153908,0.437444,2,particle
4,2025-01-01,4:00:00,eq_5,chamber_2,w_4,p-type,3,968.690498,52.487699,0.649097,2335.106388,423.234943,2.860362,0,none


### 데이터 요약 정보 확인 (info())

- **데이터의 전체적인 정보를 한눈에 확인하는 것**
- 우리가 다루는 데이터의 **각 항목이 몇 개가 있는지, 어떤 종류인지** 등을 보여주는 역할

In [10]:
# 데이터 요약 정보 확인

oxide_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   process_date           1000 non-null   object 
 1   process_time           1000 non-null   object 
 2   equipment_id           1000 non-null   object 
 3   chamber_id             1000 non-null   object 
 4   wafer_id               1000 non-null   object 
 5   wafer_doping_type      1000 non-null   object 
 6   precleaning_time       1000 non-null   int64  
 7   oxidation_temperature  999 non-null    float64
 8   oxidation_time         1000 non-null   float64
 9   pressure               995 non-null    float64
 10  gas_flow_rate          1000 non-null   float64
 11  oxide_thickness        999 non-null    float64
 12  uniformity             989 non-null    float64
 13  defect_count           1000 non-null   int64  
 14  defect_type            1000 non-null   object 
dtypes: fl

### 표본 표준 편차

- 데이터가 평균으로부터 얼마나 퍼져 있는지를 나타내는 중요한 통계 지표
    - 평균이 같아도 표본 표준 편차는 다를 수 있음.

$$
표본 표준편차(s) = \sqrt{\frac{\sum(x_i - \bar{x})^2}{n-1}}
$$

- 표본 표준 편차가 클수록 데이터의 변동성이 크고, 작을수록 데이터가 평균 주변에 밀집해 있음을 의미.
    - $ x_i $ : 개별 데이터 값
    - $ \bar{x} $ : 표본의 평균
    - $ n $ : 표본 데이터의 개수

### 사분위수 (Quartile)

- 데이터를 **4개의 부분으로 나누는 기준**
- 전체 데이터를 오름차순으로 정렬했을 때 25%, 50%, 75% 위치에 해당하는 값 (**Q1, Q2, Q3**)을 말함.

#### 사분위수를 사용하는 이유

- 평균만 보는 것보다 **분포**를 정확히 알 수 있음.
- **이상치(특히 높거나 낮은 특이한 점수)** 를 찾기 쉬움.

### 통계 정보 확인

### describe() 함수

- **데이터의 전체적인 분포를 빠르게 확인하기 위한 함수**
- 각 항목에 대해 **평균(mean), 표준편차(std), 최소값(min), 최대값(max), 중앙값(50%), 사분위수(25%, 75%)** 등을 계산해줌.
- 공정 데이터를 분석할 때, **이상치를 확인하고, 데이터가 특정 범위 내에서 적절하게 분포하는지 평가하는데 필수적**

#### describe() 함수로 알 수 있는 정보

count (데이터 개수), mean (평균값), std (표본 표준편차), min (최소값), 25% (1사분위수), 50% (중앙값, 2사분위수), 75% (3사분위수), max (최대값)

In [12]:
# 데이터 통계 정보 확인

# 수치형 데이터 통계 정보 출력
oxide_df.describe()

,precleaning_time,oxidation_temperature,oxidation_time,pressure,gas_flow_rate,oxide_thickness,uniformity,defect_count
count,1000.000000,999.000000,1000.000000,995.000000,1000.000000,999.000000,989.000000,1000.000000
mean,3.047000,998.048988,35.469981,0.547917,2996.477281,312.046403,2.593396,0.504000
std,1.410951,48.917991,20.733476,0.258748,1158.962097,91.634887,1.669576,0.920234
min,1.000000,826.428505,-32.619010,0.100204,1015.673500,125.854317,0.000480,0.000000
25%,2.000000,964.714274,20.464978,0.328828,1981.018442,237.541857,1.126819,0.000000
50%,3.000000,998.767136,35.896619,0.535148,3012.191870,310.517811,2.560923,0.000000
75%,4.000000,1032.145373,50.262327,0.774250,3992.722743,390.336687,4.026533,1.000000
max,5.000000,1153.097348,95.790972,0.998914,4997.842724,650.548250,5.500652,6.000000


## 공정 능력 분석

### 산화 공정 데이터 프레임 생성

In [13]:
# 산화 공정 데이터 읽어서 화면에 출력

import pandas as pd

file_path = "02_oxide.csv"

oxide_df = pd.read_csv(file_path)

oxide_df

,process_date,process_time,equipment_id,chamber_id,wafer_id,wafer_doping_type,precleaning_time,oxidation_temperature,oxidation_time,pressure,gas_flow_rate,oxide_thickness,uniformity,defect_count,defect_type
0,2025-01-01,0:00:00,eq_1,chamber_1,w_0,p-type,3,908.332942,30.450712,0.725458,4007.192654,212.318551,4.158928,0,none
1,2025-01-01,1:00:00,eq_2,chamber_2,w_1,n-type,1,998.595807,43.926035,0.610652,3273.074465,359.152236,3.920767,2,particle
2,2025-01-01,2:00:00,eq_3,chamber_3,w_2,p-type,5,1016.761934,66.715328,0.816619,2098.663479,442.100760,2.428848,1,pinhole
3,2025-01-01,3:00:00,eq_4,chamber_1,w_3,n-type,3,952.625393,46.845448,0.761523,2897.739657,225.153908,0.437444,2,particle
4,2025-01-01,4:00:00,eq_5,chamber_2,w_4,p-type,3,968.690498,52.487699,0.649097,2335.106388,423.234943,2.860362,0,none
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,2025-02-11,11:00:00,eq_1,chamber_3,w_995,n-type,4,1068.991935,52.783496,0.302645,4802.014246,464.065417,4.914362,0,none
996,2025-02-11,12:00:00,eq_2,chamber_1,w_996,p-type,5,976.739157,55.965298,0.696086,2534.436161,261.114871,3.471068,0,none
997,2025-02-11,13:00:00,eq_3,chamber_2,w_997,n-type,3,985.560235,45.612643,0.648714,2697.488104,310.582879,0.294383,0,none
998,2025-02-11,14:00:00,eq_4,chamber_3,w_998,p-type,2,946.535657,6.432315,0.923589,3858.784356,179.629674,2.242628,4,particle


### 목표 값 및 허용 범위

#### 산화막 두께 (웨이퍼 표면에 생성된 보호막 두께)

- 너무 얇으면 절연 효과가 부족하고, 너무 두꺼우면 성능이 저하될 수 있음.
- 따라서, **적절한 두께 범위를 유지하는 것**이 중요


#### 산화막 두께 목표값과 허용 범위($\pm 3\sigma$ 적용)

- 목표 값 = **산화막 두께의 중앙값**
- 허용 (정상)범위 = 목표값 $\pm$ 3 x 표준편차($\sigma$)

- $\pm 3\sigma$(표준 편차)를 적용하는 이유?
    - 통계적으로 $\pm 3\sigma$ 범위 안에 전체 데이터의 약 99.7%가 포함됨.
    - **데이터의 변동성이 매우 작고 예측 가능하다.**

In [14]:
# 목표값과 허용범위 계산

# 1. 산화막 두께 데이터 가져오기
oxide_thickness = oxide_df['oxide_thickness']  # 산화막 두께 컬럼 선택

# 2. 목표 값 설정을 위한 통계 계산
# 중앙값은 전체 데이터를 크기 순으로 나열했을 때 딱 가운데 위치하는 값
# 평균은 너무 크거나 작은 특이한 값의 영향을 받지만, 중앙값은 실제로 가장 많이 나오는 현실적인 값
target_thickness = oxide_thickness.median()

# 표준편차 계산
std_thickness = oxide_thickness.std()
print(f"목표값: {target_thickness:.2f}")
print(f"표준편차: {std_thickness:.2f}")

# 3. 목표값 및 허용 범위 설정 (±3σ 적용)
lower_limit = target_thickness - 3 * std_thickness  # 하한 값 (목표 값 - 3σ)
upper_limit = target_thickness + 3 * std_thickness  # 상한 값 (목표 값 + 3σ)

# 4. 결과 출력
print(f"허용 범위: {lower_limit:.2f} ~ {upper_limit:.2f}")

목표값: 310.52
표준편차: 91.63
허용 범위: 35.61 ~ 585.42


## Cp, Cpk

### Cp (공정 능력 지수, Process Capability Index)

- 공정이 규격 내에서 얼마나 균일하게 생산되는지를 나타냄.

$$
C_p = \frac{\text{USL} - \text{LSL}}{6\sigma}
$$

- **σ(시그마, 표본 표준편차)** : 제품 두께의 변동성 (즉, 제품 두께가 얼마나 들쭉날쭉한지)
    - 표본 표준편차(σ) = 샘플의 산화막 두께 변동성 $\approx$ 91.5890 $\text{\AA}$

- **USL**(Upper Specification Limit): 더 두꺼워지면 불량이 되는 최대 두께, 허용 최대 두께 (**400Å로 가정**)
- **LSL**(Lower Specification Limit): 더 얇아지면 제품이 불량이 되는 최소 두께, 허용 최소 두께 (**220Å로 가정**)

- Cp가 크면 일관성이 좋고, Cp가 작으면 크기 편차가 커서 불량이 발생할 확률이 높음.
    - Cp가 **크다** = 반도체 웨이퍼의 두께가 거의 일정 - 불량이 적고, 공정이 안정적
    - Cp가 **작다** = 웨이퍼 두께가 들쭉날쭉함 - 불량이 증가하고, 품질이 낮아짐.

    |**Cp 값**|**해석**|
    |---|---|
    |Cp < 1.00|공정의 변동이 커서 허용 범위를 벗어날 가능성이 높음 (불량률 높음)|
    |Cp = 1.00|공정이 허용 범위 내에 있지만, 변동성이 커서 품질이 불안정함|
    |Cp ≥ 1.33|공정이 안정적이며, 대체로 양품이 많음 (일반적으로 **Cp ≥ 1.33** 이상이 요구됨)|
    |Cp ≥ 2.00|매우 좋은 공정으로, 허용 범위 내에서 안정적으로 운영됨|

In [15]:
# Cp 계산

# numpy는 수학적 계산을 쉽게 할 수 있도록 도와주는 라이브러리
# 평균(mean), 표준 편차(std) 같은 통계 값을 계산할 때 사용
import numpy as np

oxide_thickness = oxide_df['oxide_thickness']

std_thickness = oxide_df['oxide_thickness'].std()

# 공정에서 허용되는 산화막 두께의 범위 설정
LSL = 220
USL = 400

# 공정능력지수(cp) 계산 공식 적용
cp_value = (USL - LSL) / (6 * std_thickness)

print(f'공정능력지수 (cp): {cp_value:.2f}')

공정능력지수 (cp): 0.33


### Cpk (공정 성능 지수, Process Capability Performance Index)

- **제품이 허용범위 중앙에 잘 맞춰져 있는지(한쪽으로 치우치지 않았는지)**를 나타냄.

$$
C_{pk} = \min\left( \frac{\text{USL} - \mu}{3\sigma}, \frac{\mu - \text{LSL}}{3\sigma} \right)
$$

- **μ (뮤, 평균 산화막 두께)**: 산화막 두께의 평균
    - 평균 두께(μ) = 산화막 두께 $\approx$ 312.0464Å
    - 평균(μ)을 사용하는 이유는 평균과 비슷한 값을 가진 산화막 두께가 가장 많기 때문에 대표값으로 평균을 사용함.

- Cpk 값이 크면 클수록 공정이 안정적이고 제품의 품질이 좋음.
    |**Cpk 값**|**해석**|
    |---|---|
    |Cpk < 1.00|공정이 한쪽으로 치우쳐 있으며, 불량이 발생할 가능성이 높음|
    |Cpk ≥ 1.33|공정이 목표값(중앙)에 잘 정렬되어 있으며, 안정적임|  

<br>

- 즉, 공정에서는 **평균을 허용범위 중앙에 최대한 정확히 맞추고, 제품의 변동성(표준편차)을 작게 유지하는 것이 가장 중요**하다 !

In [16]:
# Cpk 계산

# 평균 및 표준편차 계산
mean_thickness = oxide_thickness.mean()
std_thickness = oxide_thickness.std()

# 허용 범위
LSL = 220
USL = 400

# Cpk 계산
cpk_lower = (mean_thickness - LSL) / (3 * std_thickness)
cpk_upper = (USL - mean_thickness) / (3 * std_thickness)
cpk_value = min(cpk_lower, cpk_upper)

print(f"공정 성능 지수 Cpk: {cpk_value:.2f}")

공정 성능 지수 Cpk: 0.32


## Pandas 데이터 다루기

### 열 선택

#### Pandas에서 열 선택의 기본 형식

|**구분**|**형식**|**예시**|
|---|---|---|
|하나의 열 선택|데이터프레임['열이름']|df['이름']|
|두 개 이상 열 선택|데이터프레임[['열이름1', '열이름2']]|df['이름', '점수']|

In [23]:
# 예제

import pandas as pd

data = {
    '이름': ['철수', '영희', '민수'],
    '나이': [25, 30, 22],
    '성별': ['남자', '여자', '남자'],
    '점수': [90, 85, 95]
}

df = pd.DataFrame(data)

print(df)
print()


# 하나의 열 선택하기

# '이름'이라는 열만 선택하기
name_column = df['이름']

print(name_column)
print()


# 두 개 이상의 열 선택하기
# 두 개 이상의 열을 선택하려면 리스트([]) 안에 열 이름을 넣어야 함.

# '이름'과 '점수' 두 개의 열 선택하기
name_score_columns = df[['이름', '점수']]

print(name_score_columns)
print()


# 열 선택 후 특정 값만 보기

# 점수 열 선택 후 첫 번째 사람의 점수 보기
first_score = df['점수'][0]

print(first_score)
print()


# 여러 열 선택 후 일부 행(줄)만 보기

# '이름', '나이' 열을 선택하고 두 번째 사람 정보만 보기
second_person = df[['이름', '나이']].iloc[1]

print(second_person)

   이름  나이  성별  점수
0  철수  25  남자  90
1  영희  30  여자  85
2  민수  22  남자  95

0    철수
1    영희
2    민수
Name: 이름, dtype: object

   이름  점수
0  철수  90
1  영희  85
2  민수  95

90

이름    영희
나이    30
Name: 1, dtype: object


### 행과 열 선택

#### Pandas의 loc 함수

- 표에서 내가 보고 싶은 행과 열을 **이름으로 선택**할 때 사용하는 기능


**loc 함수의 형식**
```python
데이터프레임.loc[행, 열]
```

- 형식 예시:
    |**형식**|**의미**|
    |---|---|
    |df.loc['A', '이름']|A행의 '이름'열 선택 (값 하나)|
    |df.loc['A', :]|A행의 모든 열 선택|
    |df.loc[:, '이름']|모든 행의 '이름'열 선택|
    |df.loc['A':'C', '이름':'점수']|A부터 C행까지, '이름'부터 '점수'열까지 범위 선택|
    
    <br>
    - 콜론(:)은 모든 범위를 의미

In [28]:
# 예제

import pandas as pd

data = {
    '이름': ['철수', '영희', '민수'],
    '나이': [25, 30, 22],
    '성별': ['남자', '여자', '남자'],
    '점수': [90, 85, 95]
}

# 각 행에 이름 붙이기
df = pd.DataFrame(data, index=['A', 'B', 'C'])
print(df)
print()


# 한 개의 행 선택하기

# A행의 모든 데이터 선택하기
row_a = df.loc['A', :]
print(row_a)
print()


# 하나의 열 전체 선택하기

# '점수' 열의 모든 사람들의 점수 보기
scores = df.loc[:, '점수']
print(scores)
print()


# 특정 행과 특정 열에서 값 하나만 가져오기

# 표에서 'B' 행(영희)의 '나이' 열만 선택하기
age_b = df.loc['B', '나이']
print(age_b)
print()


# 여러 행과 여러 열을 동시에 선택하기

# A와 C의 이름과 점수만 보고 싶을 때
selected_df = df.loc[['A', 'C'], ['이름', '점수']]
print(selected_df)
print()


# 행과 열을 범위로 선택하기

# 행 A부터 C까지, 열은 나이부터 점수까지 선택하기
subset = df.loc['A':'C', '나이':'점수']
print(subset)

   이름  나이  성별  점수
A  철수  25  남자  90
B  영희  30  여자  85
C  민수  22  남자  95

이름    철수
나이    25
성별    남자
점수    90
Name: A, dtype: object

A    90
B    85
C    95
Name: 점수, dtype: int64

30

   이름  점수
A  철수  90
C  민수  95

   나이  성별  점수
A  25  남자  90
B  30  여자  85
C  22  남자  95


### 조건 검색

- 표에서 내가 원하는 조건을 만족하는 행(줄)을 골라서 찾는 것

#### 조건 검색의 구조

```python
데이터프레임[데이터프레임['열 이름'] 조건]
```

- 데이터 조건검색의 다양한 조건 :
    - 같다, 크다, 작다, 이상, 이하, 다르다

In [35]:
# 예제

import pandas as pd

data = {
    '이름': ['철수', '영희', '민수', '수진'],
    '나이': [25, 30, 22, 28],
    '성별': ['남자', '여자', '남자', '여자'],
    '점수': [90, 85, 95, 88]
}

df = pd.DataFrame(data)
print(df)
print()


# 간단한 조건 검색 해보기

# 성별이 여자인 사람만 고르기
female_df = df[df['성별'] == '여자']
print(female_df)
print()

# 나이가 25살이 넘는 사람 고르기
age_over_25 = df[df['나이'] > 25]
print(age_over_25)
print()


# 두 가지 조건 사용하기
# 조건이 2개 이상일 때는 각각 괄호로 묶어야 함.
# AND 조건(모두 만족)은 '&' 사용
# OR 조건(하나만 만족)은 '|' 사용

# 성별이 여자인 사람 & 점수가 85점 이상인 사람만 뽑기
female_high_score = df[(df['성별'] == '여자') & (df['점수'] >= 85)]
print(female_high_score)
print()

# 나이가 30살 이상이거나 점수가 90점 넘는 사람 뽑기
age_or_score = df[(df['나이'] >= 30) | (df['점수'] >= 90)]
print(age_or_score)

   이름  나이  성별  점수
0  철수  25  남자  90
1  영희  30  여자  85
2  민수  22  남자  95
3  수진  28  여자  88

   이름  나이  성별  점수
1  영희  30  여자  85
3  수진  28  여자  88

   이름  나이  성별  점수
1  영희  30  여자  85
3  수진  28  여자  88

   이름  나이  성별  점수
1  영희  30  여자  85
3  수진  28  여자  88

   이름  나이  성별  점수
0  철수  25  남자  90
1  영희  30  여자  85
2  민수  22  남자  95
